src/lexer/Lexer.ts

```diff
@@ -13,474 +15,504 @@ import {
     Double
 } from "../brsTypes";
 
-/** The zero-indexed position at which the token under consideration begins. */
-let start: number;
-/** The zero-indexed position being examined for the token under consideration. */
-let current: number;
-/** The one-indexed line number being parsed. */
-let line: number;
-
-/** The BrightScript code being converted to an array of `Token`s. */
-let source: string;
-/** The tokens produced from `source`. */
-let tokens: Token[];
-
-/**
- * Converts a string containing BrightScript code to an array of `Token` objects that will later be
- * used to build an abstract syntax tree.
- *
- * @param toScan the BrightScript code to convert into tokens
- * @returns a (read-only) array of tokens to be passed to a parser.
- */
-export function scan(toScan: string): ReadonlyArray<Token> {
-    start = 0;
-    current = 0;
-    line = 1;
-
-    source = toScan;
-    tokens = [];
-
-    while (!isAtEnd()) {
-        start = current;
-        scanToken();
+/** The results of a Lexer's scanning pass. */
+interface ScanResults {
+    /** The tokens produced by the Lexer. */
+    tokens: ReadonlyArray<Token>,
+    /** The errors encountered by the Lexer. */
+    errors: ReadonlyArray<BrsError>
+}
+
+export class Lexer {
+    /** Allows consumers to observe errors as they're detected. */
+    readonly events = new EventEmitter();
```

Comment:
Rather than needing to call `myLexer.events.on` where it doesn't tell you what events are exposed, you should consider creating your own "on" method that wraps the events internally, so that consumers of this class get better intellisense as to what actually is acceptable. Same concept applies for the Parser too. Something like this:

```typescript
class Lexer{
    private eventEmitter = new EventEmitter();

    public on(eventName: 'error', callback: (err: BrsError) => void);
    public on(eventName: string, callback: (data: any) => void) {
        this.eventEmitter.on(eventName, callback);
        //create a disconnect function that the caller can use to stop listening to this event
        let disconnect = ()=>{
            this.eventEmitter.removeListener(eventName, callback);
        };
        return disconnect;
    }
}
```

Reply:
I spent about an hour trying to find a typesafe way to do this, trying to avoid writing custom wrappers. Writing those wrappers would have taken ten minutes. 😐

I’ll add one before I merge (and will likely make `.events` private).